src/primitives/sinsemilla.rs

```diff
@@ -64,59 +61,107 @@ impl<I: Iterator<Item = bool>> Iterator for Pad<I> {
     }
 }
 
-#[allow(non_snake_case)]
-fn Q(domain_prefix: &str) -> pallas::Point {
-    pallas::Point::hash_to_curve(GROUP_HASH_Q)(domain_prefix.as_bytes())
+/// Trait representing a domain in which [`sinsemilla::hash_to_point`] and
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub trait HashDomain {
+    #[allow(non_snake_case)]
+    fn Q(&self) -> pallas::Point;
+}
+
+/// A domain with a specific prefix string, in which [`sinsemilla::hash_to_point`] and
+/// [`sinsemilla::hash`] can be used.
+///
+/// [`sinsemilla::hash_to_point`]: self::hash_to_point
+/// [`sinsemilla::hash`]: self::hash
+pub struct Domain(pub &'static str);
+
+impl HashDomain for Domain {
+    fn Q(&self) -> pallas::Point {
+        pallas::Point::hash_to_curve(Q_PERSONALIZATION)(self.0.as_bytes())
```

Comment:
I suggest doing the `hash_to_curve` only when constructing a domain, not later (we only need to access Q on a domain, not the original string).